### PR TITLE
Reorganize AES implementation internals

### PR DESCRIPTION
**Pull request by Rocknest**
- group functions logically
- move constants to one place, name them descriptively 
- more type safe code
- more generic code

I would like to change AES public api a bit so there might be a part2.

### CODE REVIEW

File: lib/std/crypto/aes.zig [hunk 1]
```diff
@@ -9,216 +9,262 @@ const std = @import("../std.zig");
...
+        fn subw(w: u32) u32 {
+            return @as(u32, sbox0[w >> 24]) << 24
+                | @as(u32, sbox0[w >> 16 & 0xff]) << 16
+                | @as(u32, sbox0[w >> 8 & 0xff]) << 8
+                | @as(u32, sbox0[w & 0xff]);
         }
-        pub fn ctr(ctx: Self, dst: []u8, src: []const u8, iv: [16]u8) void {
-            ctx.enc.ctr(dst, src, iv);
+
+        fn rotw(w: u32) u32 {
```
Comment (daurnimator) [inline lib/std/crypto/aes.zig:160]: keep this out of the struct.. or actually; this should be replaced with `std.math.rotl(u32, w, 8)`

File: lib/std/crypto/aes.zig [hunk 2]
```diff
@@ -9,216 +9,262 @@ const std = @import("../std.zig");
...
-            var dec: Decrypt = undefined;
-            expandKeyDecrypt(ctx.enc[0..], dec.dec[0..]);
-            return dec;
+        pub fn encrypt(ctx: Self, dst: []u8, src: []const u8) void {
+            AES.encryptBlock(ctx.expkey, src[0..16], dst[0..16]);
         }
 
-        pub fn encrypt(ctx: Self, dst: []u8, src: []const u8) void {
-            encryptBlock(ctx.enc[0..], dst, src);
+        fn xorBytes(noalias a: *const [16]u8, b: *const [16]u8, dst: *[16]u8) void {
```
[the remark this answers is not in the document]
Comment (Rocknest) [reply lib/std/crypto/aes.zig:218]: I will move it with `ctr` into another file in an upcoming PR

**Comment by jedisct1** [comment]
This is cleaner, thanks!

And moving `ctr` elsewhere is a good thing, as AES-CTR should share the same API as other stream ciphers rather than being a thing of its own. Keeping all AES-based constructions in the same file also wouldn't scale.

Our AES API still has some design issues limiting its use cases. In particular, the round function is not exposed, which prevents it from being usable to implement AES-PRF, AEGIS, and other recent AES-based construction that don't use the original key schedule.

I have preliminary code that addresses this and adds AES-NI and wide block support. This also requires some API changes, but these can fit nicely on top of yours.

**Comment by jedisct1** [comment]
The reorganization was done in #6398, that includes these diffs, along with the proposed future changes.

This issue can probably be closed now.

Thanks a lot for suggesting these changes! AES was indeed in need of a more versatile and more organized API.

**Pull request closed without merging**